rfcs/0120-artifact-service.md

```diff
@@ -0,0 +1,296 @@
+  frontend service and objects, with JWT being a possibility
+* The frontend service will block new artifacts after task resolution by not issuing new
+  signed urls to this service
+* Data storage in objects will be done with Postgres and not Azure Table
+  Storage
+* We do not intend to have an operation to list objects in the API
+* The redirecting logic of the object service will adhere to the HTTP spec
+  using 300-series redirects where possible
+* Where content-negotiation is impossible, the service will block retreival or
+  creation of objects when incoming requests aren't compatible with required
+  response
+* We will support uploading objects to different regions and services so that
+  we can minimize interregion transfer costs for initial object creation
+
+### Content Negitiation
+Some storage providers allow for a specific content-encoding to be set for
+their resources, but do not allow for content negotiation.  An example is S3,
+where you can store objects with an immutable content-encoding value of 'gzip'
+at creation.  All requests will respond with "Content-Encoding: gzip"
+regardless of the "Accept-Encoding" request header value.  The HTTP spec is not
+100% clear on whether or not this is allowed per the spec, so we will ensure
+that only defintely valid requests are processed.
+
+In order to force only valid requests, we will require that any request which
+would result in a forced content-encoding header have a corresponding
+accept-encoding header.  If a request which would be forced to respond with a
+"Content-Encoding: gzip" then only requests with "Accept-Encoding: gzip" (or
+compatible) would succeed.  In this case a 406 Not Acceptable response would be
+given.
+
+This does not implement or intend to implement content-negotiation, instead is
+a system to ensure that only requests which would've negotiated a compatible
+encoding would succeed.
+
+## API
+This API is up for debate and is only an illustrative starting point.  
+
+### Generalised request format
+There will be a concept of a generalised HTTP request.  These requests are used
+to transmit metadata of an HTTP request which the  service requires the client
+to perform.
+
+An example usage of this format is uploading a file.  The creation endpoint of
+this service will generate all of the HTTP request metadata that the uploading
+process will use to upload file.  Then the uploader will match these HTTP
+requests with payload bodies and run each of these requests.
+
+The motivation for this type of flow is to support the case where the uploader
+is running in a different region than this service is.  If we required all
+uploads to gate through this service, we'd need to duplicate uploading effort
+and perform unneeded interregion transfers.  By transmitting a list of requests
+which allow the uploading to happen, we can instruct the uploader to upload
+directly to the existing S3 infrastructure, but perform all signing within this
+service.
```

Comment:
Ah, this is the information I was missing earlier on - maybe we just need to present this earlier in the RFC, to provide context for the references to signed URLs.